packages/runtime/container-runtime/src/dataStore.ts

```diff
@@ -0,0 +1,187 @@
+        && typeof maybeDataStoreAliasMessage?.alias === "string";
+};
+
+/**
+ * Encapsulates the return codes of the aliasing API
+ */
+export enum AliasResult {
+    /**
+     * The datastore has been successfully aliased
+     */
+    Success = "Success",
+    /**
+     * There is already a datastore bound to the provided alias
+     */
+    Conflict = "Conflict",
+    /**
+     * The datastore is currently in the process of being aliased
+     */
+    Aliasing = "Aliasing",
+    /**
+     * The datastore has been attempted to be aliased before
+     */
+    AlreadyAliased = "AlreadyAliased",
+}
+
+/**
+ * A fluid router with the capability of being assigned an alias
+ */
+ export interface IDataStore extends IFluidRouter {
+    /**
+     * Attempt to assign an alias to the datastore.
+     * If the operation succeeds, the datastore can be referenced
+     * by the supplied alias.
+     *
+     * @param alias - Given alias for this datastore.
+     */
+    trySetAlias(alias: string): Promise<AliasResult>;
+}
+
+export const channelToDataStore = (
+    fluidDataStoreChannel: IFluidDataStoreChannel,
+    internalId: string,
+    runtime: ContainerRuntime,
+    datastores: DataStores,
+    logger: ITelemetryLogger,
+): IDataStore => new DataStore(fluidDataStoreChannel, internalId, runtime, datastores, logger);
+
+enum AliasState {
+    Aliased = "Aliased",
+    Aliasing = "Aliasing",
+    None = "None",
+}
+
+class DataStore implements IDataStore {
+    private aliasState: AliasState = AliasState.None;
+    private alias: string | undefined;
+
+    async trySetAlias(alias: string): Promise<AliasResult> {
+        switch (this.aliasState) {
+            // If we're already aliasing, throw an exception
+            case AliasState.Aliasing:
+                return AliasResult.Aliasing;
+            // If this datastore is already aliased, return true only if this
+            // is a repeated call for the same alias
+            case AliasState.Aliased:
+                return this.alias === alias ? AliasResult.Success : AliasResult.AlreadyAliased;
+            // There is no current or past alias operation for this datastore,
+            // it is safe to continue execution
+            case AliasState.None: break;
+            default: unreachableCase(this.aliasState);
+        }
+
+        this.aliasState = AliasState.Aliasing;
+        const message: IDataStoreAliasMessage = {
+            internalId: this.internalId,
+            alias,
+        };
+
+        if (this.runtime.attachState === AttachState.Detached) {
+            const localResult = this.datastores.processAliasMessageCore(message);
+            // Explicitly Lock-out future attempts of aliasing,
+            // regardless of result
+            this.aliasState = AliasState.Aliased;
+            return localResult ? AliasResult.Success : AliasResult.Conflict;
+        }
+
+        this.fluidDataStoreChannel.bindToContext();
```

Comment:
Feels like this should move up, above detached container case.
Do we have UT for detached container case?

Reply:
Ahh.. great call. Thank you! 👍 

Yes, we do.

Reply:
See `Assign multiple data stores to the same alias, first write wins, same container - detached` inside `packages/test/test-end-to-end-tests/src/test/rootDatastores.spec.ts`